File: ldp/utils_benchmarking_ldp.py
#confidence intervals for multinomial proportion
def mltnp_ci(tendency, cl):
    #frequency table
    ft  = np.unique(ar = tendency, 
                    return_counts = True)
    #number of observations
    n = sum(ft[1])
    #tendency indicator length (3)
    k = len(ft[1])
    #proportions
    p = ft[1] / n
    #confidence interval elements
    q_chi = stats.chi2.ppf(1 - (1 - cl) / k, df = 1)
    lci = (q_chi + 2 * ft[1] - np.sqrt(q_chi * (q_chi + 4 * ft[1] * (n - ft[1]) / n))) / (2 * (n + q_chi))
    uci = (q_chi + 2 * ft[1] + np.sqrt(q_chi * (q_chi + 4 * ft[1] * (n - ft[1]) / n))) / (2 * (n + q_chi))
    #summary table
    res = pd.DataFrame({"est": p, 
                        "lower": np.maximum(0, lci), 
                        "upper": np.minimum(1, uci)},
                       index = ft[0])
    
    return res

#average deviation confidence interval
def d_ci(deviation, cl):
    #frequency table
    ft  = np.unique(ar = deviation, 
                    return_counts = True)
    #number of obligors
    n = len(deviation)
    #deviation probabilities
    p = ft[1] / n
    #possible sums
    p_sums = np.arange(min(deviation).item() * n, max(deviation).item() * n + 1)
    #average deviation (D values)
    d_avg = p_sums / n
    #cumulative probabilities    
    cnv_res = convolve_p(p = p, 
                         n = n)
    cdf_avg = np.cumsum(cnv_res)
    #confidence level index
    idx_l = np.where(cdf_avg >= (1 - cl) / 2)[0][0]
    idx_u = np.where(cdf_avg >= (1 + cl) / 2)[0][0]
    
    #confidence interval for average deviation
    res = pd.DataFrame([{"est": np.mean(deviation), 
                        "lower": d_avg[idx_l], 
                        "upper": d_avg[idx_u]}]) 
    
    return res
#convolve n times function
def convolve_p(p, n):
    p_i = p
    for _ in range(1, n):
        p_i = np.convolve(a = p_i, 
                          v = p, 
                          mode = "full")
    return p_i

File: ldp/utils_benchmarking_ldp.R
#confidence intervals for multinomial proportion
mltnp.ci <- function(tendency, cl) {
      #frequency table
      ft <- table(tendency)
      #number of observations
      n <- sum(ft, na.rm = TRUE)
      #tendency indicator length (3)
      k <- length(ft)
      #proportions
      p <- ft / n
      #confidence interval elements
      q.chi <- qchisq(1 - (1 - cl)/k, df = 1)
      lci <- (q.chi + 2*ft - sqrt(q.chi*(q.chi + 4*ft*(n - ft)/n))) / (2*(n + q.chi))
      uci <- (q.chi + 2*ft + sqrt(q.chi*(q.chi + 4*ft*(n - ft)/n))) / (2*(n + q.chi))  
      #summary table
      res <- cbind(est = p, 
                   lower = pmax(0, lci), 
                   upper = pmin(1, uci))
return(res)
}

#average deviation confidence interval
d.ci <- function(deviation, cl) {
      #frequency table
      ft <- table(deviation)
      #number of obligors
      n <- length(deviation)
      #deviation probabilities
      p <- ft / n
      #possible sums
      p.sums <- seq(from = min(deviation)*n, 
                    to = max(deviation)*n, 
                    by = 1)
      #average deviation (D values)
      d.avg <- p.sums / n
      #cumulative probabilities
      cnv.res <- convolve.p(p = p, 
                            n = n)
      cdf.avg <- cumsum(cnv.res)
      #confidence level index
      idx.l <- which(cdf.avg >= (1 - cl) / 2)[1]
      idx.u <- which(cdf.avg >= (1 + cl) / 2)[1]
      #confidence interval for average deviation
      res <- data.frame(est = mean(deviation),
                        lower = d.avg[idx.l],
                        upper = d.avg[idx.u])
return(res)
}

#convolve n times function
convolve.p <- function(p, n) {  
      p.i <- p
      for   (i in 2:n) {
            p.i <- convolve(x = p.i, 
                            y = rev(p), 
                            type = "open")
            }
return(p.i)
}




